scoring/precision-bscore.metta

```diff
@@ -0,0 +1,351 @@
+;;
+;; The function coordinates the entire precision scoring process:
+;; 1. Compress raw data into CTable format (group identical input patterns)
+;; 2. Calculate individual precision for each compressed pattern  
+;; 3. Sort patterns by precision in descending order
+;; 4. Run greedy best_possible_bscore algorithm to find optimal score
+;;
+;; Parameters:
+;;   $rows (Expression): Raw data rows from filtered table
+;;   $labels (Expression): Column labels for the filtered table
+;;   $activationPressure (Number): Penalty scaling for activation constraints
+;;   $minActivation (Number): Minimum acceptable activation ratio (early stopping condition)
+;;   $maxActivation (Number): Maximum acceptable activation ratio  (for penalty calculation)
+;;   $totalWeight (Number): Total dataset size (for activation calculations)
+;;
+;; Returns: (Number) Best achievable precision score using these features
+;;
+;; This represents the theoretical upper bound on precision that any classifier
+;; could achieve using only the selected features while respecting activation
+;; constraints. It's used by hill climbing to evaluate feature set quality.
+(: precisionBScore (-> Expression Expression Number Number Number Number Number Number Number))
+(= (precisionBScore $rows $labels $activationPressure $minActivation $maxActivation $totalWeight)
+   (let* (
+            (() (println! (debug Inside Precision scoring)))
+            ($groupedData (cTable $rows $labels))
+            
+            ($precisionRows (createPrecisionRows $groupedData))
+            
+            ($bestPreScore (bestPossibleBScore $precisionRows 0.0 0.0 $totalWeight $minActivation $maxActivation 
+                                         $activationPressure -1.0000000000000001e308))
+         )
+     $bestPreScore
+   )
+)
+
+;; Extract selected features plus target column from full dataset
+;; Input:
+;;   $featureIndices:  selected feature column indices
+;;   $rows: Original table data
+;;   $labels: Column labels
+;;
+;; Output:
+;;   Expression like (filtered_rows filtered_labels) containing only selected features + target
+(: extractTable (-> Expression (List (List $a)) (List Symbol) (Expression Expression)))
+(= (extractTable $featureIndices $rows $labels)
+   (let* (
+          ;; Target is always last column
+          ($targetIndex (- (List.length $labels) 1))
+          ;; Combine selected features with target index
+          ($exprIndices (appendAtom $targetIndex $featureIndices))
+
+          ;; Extract selected columns from each row
+          ($filteredRows (zipColumn $exprIndices $rows))
+          ;; Extract corresponding labels
+          ($filteredLabels (map-atom $exprIndices $each (List.at $each $labels)))
+          ($filteredExpr (List.listToExpr $filteredRows))
+          )
+     ($filteredExpr $filteredLabels)
+   )
+)
+
+;; Main pre_scorer implementation 
+;; This is the top-level function called by hill climbing for feature selection
```

Comment:
As far as I know. This isn't the only scorer being called by the hill climber. It actually uses `mi` by default doesn't it? Unless the user specifically specifies the socrer to be `pre_scorer`?

Reply:
Regardless, this is still a useful addition to the current scorer we have